Pavel/9469

sagor999 opened the pull request:
## Description
<!-- Describe your changes in detail -->

## Related Issue(s)
<!-- List the issue(s) this PR solves -->
Fixes #

## How to test
<!-- Provide steps to test this PR -->

## Release Notes
<!--
  Add entries for the CHANGELOG.md or "NONE" if there aren't any user facing changes.
  Each line becomes a separate entry.
  Format: [!<optional for breaking>] <description>
  Example: !basic auth is no longer supported
  See https://www.notion.so/gitpod/Release-Notes-513a74fdd23b4cb1b3b3aefb1d34a3e0
-->
```release-note
```

## Documentation
<!--
Does this PR require updates to the documentation at www.gitpod.io/docs?
* Yes
  * 1. Please create a docs issue: https://github.com/gitpod-io/website/issues/new?labels=documentation&template=DOCS-NEW-FEATURE.yml&title=%5BDocs+-+New+Feature%5D%3A+%3Cyour+feature+name+here%3E
  * 2. Paste the link to the docs issue below this comment
* No
  * Are you sure? If so, nothing to do here.
-->

roboquat commented:
@sagor999: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

sagor999 commented:
/werft run

sagor999 commented:
/werft run with-clean-slate-deployment

sagor999 commented:
/werft run with-clean-slate-deployment

sagor999 commented:
@csweichel I believe I addressed your concerns from earlier sync call. Let me know what you think.
snapshot volumes are now tied to workspace instance id, so that it will be easier to tie them to prebuilds in the later PR. 


Also added loom video for other teams showing how this feature works.

sagor999 commented:
/werft run

sagor999 commented:
/werft run with-clean-slate-deployment=true

sagor999 commented:
Adding a loom video walking through all of my code changes explaining them, so that hopefully it gives reviewer a bit more context for the changes.
https://www.loom.com/share/46b5719f32894af7b3817d784ef6f608

jenting commented:
Since it's a quite large PR, give me one more day to review it, thx 🙏

jenting reviewed (commented):
Could we consider rename all the variable from `snapshot volume` to `volume snapshot`? 

It makes consistent with existing code and Kubernetes naming `VolumeSnapshot`/`VolumeSnapshotContent`/`VolumeSnapshotClass`.

jenting commented on components/ws-manager/go.mod:
```diff
@@ -73,6 +73,8 @@ require (
 	github.com/json-iterator/go v1.1.12 // indirect
 	github.com/klauspost/cpuid v1.3.1 // indirect
+	github.com/klauspost/pgzip v1.2.5 // indirect
+	github.com/kubernetes-csi/external-snapshotter/client/v6 v6.0.0-rc4 // indirect
```
> Also, the min Kubernetes verion for snapshotter is 1.20. What's our promised min Kubernetes version for self-hosted solution?

Okay, I found the answer https://www.gitpod.io/docs/self-hosted/latest/cluster-set-up#supported-kubernetes-distributions Our min Kubernetes version for self-hosted solution is 1.21.

sagor999 replied:
It doesn't seem like v5 is available:
```
go get -u github.com/kubernetes-csi/external-snapshotter/client/v5
go: module github.com/kubernetes-csi/external-snapshotter@upgrade found (v1.2.2), but does not contain package github.com/kubernetes-csi/external-snapshotter/client/v5
```
unless I am doing something wrong. :thinking: 
Though `v4` works, so I will switch to that one.

jenting replied:
Let me check 👨‍💻 
Since we install snapshot-controller v5.0.1, it'd be better we use the same version 😄

jenting replied:
Okay, the problem is the client/v5.0.0 released with v4 version https://github.com/kubernetes-csi/external-snapshotter/blob/client/v5.0.0/client/go.mod#L1. Ref to https://github.com/kubernetes-csi/external-snapshotter/pull/670#discussion_r833332402

I think we should downgrade the snapshot controller and the CRD version to v4.2.x (either v4.2.0 or v4.2.1) in workspace repo.

sagor999 replied:
I think version of the client doesn't matter here. Why?
We are using GA version of the CRD, and I use the client here only as a way to interact with CRD easier (so that I don't need to use UnstructuredData to create\update k8s object). 
We don't use client for anything else.

jenting replied:
I see.

I check the kubernetes-csi/external-snapshotter releases, looks like v4 is the main focus released (having lots of major/minor release). I feel like v5 is an intermediate release. Do you think we should keep the snapshot-controller as it is _or_ keep the snapshot-controller major/minor version matches to the client version?

jenting replied:
Okay, let's keep is as is.

sagor999 commented on components/ws-manager/cmd/run.go:
```diff
@@ -32,6 +32,7 @@ import (
 	imgbldr "github.com/gitpod-io/gitpod/image-builder/api"
 	"github.com/gitpod-io/gitpod/ws-manager/pkg/manager"
 	"github.com/gitpod-io/gitpod/ws-manager/pkg/proxy"
+	volumesnapshotv1 "github.com/kubernetes-csi/external-snapshotter/client/v6/apis/volumesnapshot/v1"
```
For workspace preview we install it from here:
https://github.com/gitpod-io/workspace-preview/tree/main/gcp/deploy

For prod I believe we are only installing GCP CSI driver only for now.

For dev env - we need to engage platform team to see how they deploy gitpod in there.

sagor999 commented on components/ws-manager/pkg/manager/create.go:
```diff
@@ -252,7 +253,18 @@ func (m *Manager) createPVCForWorkspacePod(startContext *startWorkspaceContext)
-	}, nil
+	}
+
+	if startContext.VolumeSnapshot.PvcSnapshotVolumeName != "" {
```
This will be done in a separate PR. What will be done is that we would need to check if that volume snapshot is present in the cluster. If not, then create it and use `handle` to link it to correct snapshot in gcp project.

sagor999 commented on components/ws-manager/pkg/manager/monitor.go:
```diff
@@ -927,39 +928,156 @@ func (m *Monitor) finalizeWorkspaceContent(ctx context.Context, wso *workspaceOb
 				}
+				createdSnapshotVolume = true
 			}
+			if createdSnapshotVolume {
```
Those are needed, because `doFinalize := func()` may be called multiple times. So we need to make sure we can resume from the right spot.

sagor999 commented on components/ws-manager/pkg/manager/monitor.go:
```diff
@@ -927,39 +928,156 @@ func (m *Monitor) finalizeWorkspaceContent(ctx context.Context, wso *workspaceOb
-				err = m.manager.markWorkspace(ctx, workspaceID, addMark(workspaceExplicitFailAnnotation, err.Error()))
+		if pvcFeatureEnabled {
+			pvcName := wso.Pod.Name
+			if !createdSnapshotVolume {
```
see [1]

sagor999 commented on components/ws-manager/pkg/manager/monitor.go:
```diff
@@ -927,39 +928,156 @@ func (m *Monitor) finalizeWorkspaceContent(ctx context.Context, wso *workspaceOb
+			}
+
+			// backup is done and we are ready to kill the pod, mark PVC for deletion
+			if readySnapshotVolume && !deletedPVC {
```
see [1]

jenting commented on components/ws-manager/pkg/manager/monitor.go:
```diff
@@ -927,39 +928,156 @@ func (m *Monitor) finalizeWorkspaceContent(ctx context.Context, wso *workspaceOb
+					},
+				}
+
+				err = m.manager.Clientset.Create(ctx, volumeSnapshot)
```
I agree that it should never happen, but it still have the corner case that make it probably happen.

How about we check the error is already existed error and skip it (don't error out), and keep the run the code flow `createdVolumeSnapshot=true`. That is
```go
...
    err = m.manager.Clientset.Create(ctx, volumeSnapshot)
    if err != nil && !k8serr.IsAlreadyExists(err) {
        err = xerrors.Errorf("cannot create volumesnapshot: %v", err)
        return true, nil, err
    }

    createdVolumeSnapshot = true
}
if createdVolumeSnapshot {
    ....
}
```

sagor999 replied:
Good point! Will make the change.

jenting commented on components/ws-manager/pkg/manager/monitor.go:
```diff
@@ -927,7 +928,122 @@ func (m *Monitor) finalizeWorkspaceContent(ctx context.Context, wso *workspaceOb
+						return false, err
+					}
+					if volumeSnapshot.Status != nil {
+						if volumeSnapshot.Status.ReadyToUse != nil && *(volumeSnapshot.Status.ReadyToUse) {
```
nit, personally I'd prefer to use a function to do check (is not nil and is true/false), refer to [code](https://github.com/vmware-tanzu/velero/blob/e9c997839ebbb04688a88ffb981eeb74de51b380/pkg/util/boolptr/boolptr.go#L19-L27)

I've check the code in the common-g, looks like currently we don't have the same function. We could enhance it in the future 🛹

jenting commented on components/ws-manager/pkg/manager/monitor.go:
```diff
@@ -927,7 +928,122 @@ func (m *Monitor) finalizeWorkspaceContent(ctx context.Context, wso *workspaceOb
+					}
+					if volumeSnapshot.Status != nil {
+						if volumeSnapshot.Status.ReadyToUse != nil && *(volumeSnapshot.Status.ReadyToUse) {
+							pvcVolumeSnapshotContentName = *volumeSnapshot.Status.BoundVolumeSnapshotContentName
```
In general, the BoundVolumeSnapshotContentName should not be nil.
But it'd be better check the nil first because there probably the `status.ReadyToUse == true` but `status.BoundVolumeSnapshotContentName == nil`.

jenting commented on components/ws-manager/pkg/manager/monitor.go:
```diff
@@ -927,7 +928,122 @@ func (m *Monitor) finalizeWorkspaceContent(ctx context.Context, wso *workspaceOb
+				readyVolumeSnapshot = true
+			}
+			if readyVolumeSnapshot && !markVolumeSnapshotAnnotation {
+				var volumeSnapshotContent volumesnapshotv1.VolumeSnapshotContent
```
We could add the VolumeSnapshotContentName in the log at this entrypoint, so the following logs, we don't have to add `.WithField("VolumeSnapshotContent.Name", pvcVolumeSnapshotContentName)`
```go
log = log.WithField("VolumeSnapshotContent.Name", pvcVolumeSnapshotContentName)
```

jenting commented on components/ws-manager/pkg/manager/monitor.go:
```diff
@@ -927,7 +928,122 @@ func (m *Monitor) finalizeWorkspaceContent(ctx context.Context, wso *workspaceOb
+						},
+					},
+				)
+				if pvcErr != nil {
```
```suggestion
                                if pvcErr != nil && !k8serr.IsNotFound(pvcErr) {
```

jenting reviewed (commented):
Recently PRs deprecated the `xerrors.Errorf`, and use `fmt.Errorf` instead.
We could update it if you want _or_ in the following PR.

sagor999 commented on components/ws-manager/pkg/manager/monitor.go:
```diff
@@ -927,7 +928,122 @@ func (m *Monitor) finalizeWorkspaceContent(ctx context.Context, wso *workspaceOb
+						Name:      pvcVolumeSnapshotName,
+						Namespace: m.manager.Config.Namespace,
+						Labels: map[string]string{
+							"workspaceID": workspaceID,
```
Ah, that was a test to see if possible to propagate volume snapshot labels to GCP snapshot label. But that did not work (have a separate task to investigate if possible to do that). 
So I will remove the label for now.

jenting replied:
Roget that, I saw it's in epic #7901 already.

jenting replied:
Related issue #10186

sagor999 commented:
@jenting PTAL :heart:

jenting reviewed (approved):
Did not launch a cluster to test (will do it tomorrow) but workspace related code LGTM.
Let's ship the 🛹 , thank you @sagor999 for the hard work.

corneliusludmann reviewed (approved):
Mainly focused on reviewing the self-hosted team owned files. These are looking awesome. :man_dancing: Ship it! :ship:

sagor999 commented:
@aledbf can you please review `/components/ws-manager-api` as you are code owner of that component.
@Furisto can you please review '/components/ws-daemon-api`, it is just formatting changes.
@geropl can you please review `initializer.proto` in `/components/content-service-api`. You are the owner of that component. And it just one line change. 

Thank you!

jenting commented on components/ws-manager/pkg/manager/create.go:
```diff
@@ -256,10 +257,19 @@ func (m *Manager) createPVCForWorkspacePod(startContext *startWorkspaceContext)
 	}
 
-	return pvc, nil
+	if startContext.VolumeSnapshot != nil {
```
I make a test a workspace-preview, I can't start a new workspace.

I trace the log and code, looks like we need to check if `startContext.VolumeSnapshot.VolumeSnapshotName != "'` as well.

```go
if startContext.VolumeSnapshot != nil && startContext.VolumeSnapshot.VolumeSnapshotName != "" {
```

sagor999 replied:
Indeed! Pushed fix. Thank you!

sagor999 commented:
/hold
to look at the issue that JenTing mentioned

sagor999 commented:
/unhold
verified, works now.
